Update PNG rendering

mauch opened the pull request:
Use the zscale algorithm from katsdpsigproc now that it has moved there. The rest is largely a cut-and-paste job from `render.py` in katsdpimager. The commit messages describe the changes I needed to make for continuum images.

@bmerry, as we discussed, this (or a variant) could go somewhere else where both katsdpimager and katsdpcontim can import it. I suspect zscale could also leave katsdpsigproc and join it as well. As such this PR is really just an indication of how things will work for the continuum case so we can identify whether or not the implementations have enough in common to move things somewhere else. I'll put a 'do not merge' on this while we work that out.

bmerry commented on katacomb/requirements.txt:
```diff
@@ -43,9 +42,11 @@ requests
-katsdpservices @ git+https://github.com/ska-sa/katsdpservices
+katsdpsigproc @ git+https://github.com/ska-sa/katsdpsigproc
 katsdptelstate @ git+https://github.com/ska-sa/katsdptelstate
+katsdpservices @ git+https://github.com/ska-sa/katsdpservices
```
Why did this move below telstate? They were alphabetical before.

mauch replied:
Whoops! I fixed this is the latest commit.

Assuming this passes the tests you can take a look now.

bmerry commented on katacomb/katacomb/util/image_util.py:
```diff
@@ -1,215 +1,94 @@
-    scaling.
+use('Agg', warn=False)  # noqa: E402
+import matplotlib.axes
+from matplotlib import pylab as plt
```
pylab [isn't recommended](https://matplotlib.org/api/pyplot_api.html) for programmatic use.
```suggestion
import matplotlib.pyplot as plt
```

bmerry commented on katacomb/katacomb/util/image_util.py:
```diff
@@ -1,215 +1,94 @@
+    slices : list
+        Choice of image dimensions. Passed to :class:`WCSAxes`
+    trim : float
+        Fraction of pixels on each edge of the image to clip
```
Have you considered automatically trimming to the bounding box of non-NaN pixels?

mauch replied:
That's a good idea.

I'll have to think about / experiment with the best way to do it.

bmerry replied:
I imagine something involving np.min/np.max, np.nonzero and np.isfinite should be able to give axis bounds fairly easily. It will just need some corner case handling in case an image is all-NaN for some reason.

bmerry commented on katacomb/katacomb/util/image_util.py:
```diff
@@ -88,7 +86,15 @@ def write_image(input_file, output_file, width=1024, height=768, dpi=DEFAULT_DPI
+        finite_data = np.where(np.isfinite(data))
+        bbox = (0, image_width, 0, image_height)
+        if data[finite_data].size > 0:
+            ymax, xmax = np.max(finite_data, axis=1)
+            ymin, xmin = np.min(finite_data, axis=1)
```
This works but is going to create new temporary arrays that are potentially hundreds of megabytes (because `finite_data` is a tuple of two arrays, not an array itself). I'd suggest
```suggestion
            ymin = np.min(finite_data[0])
            ymax = np.max(finite_data[0])
            xmin = np.min(finite_data[1])
            xmax = np.max(finite_data[1])
```

bmerry commented on katacomb/katacomb/util/image_util.py:
```diff
@@ -88,7 +86,15 @@ def write_image(input_file, output_file, width=1024, height=768, dpi=DEFAULT_DPI
+        # Plot the lot if any axis is completely blanked
+        finite_data = np.where(np.isfinite(data))
+        bbox = (0, image_width, 0, image_height)
+        if data[finite_data].size > 0:
```
That's quite an expensive way to check if there are any finite pixels - it's going to create a new array with all the finite pixel values. I'd suggest
```suggestion
        if finite_data[0].size > 0:
```

bmerry commented on katacomb/katacomb/util/image_util.py:
```diff
@@ -4,23 +4,23 @@
-    x_trim, y_trim = image_width * trim, image_height * trim
-    ax.set_xlim(-0.5 + x_trim, image_width - x_trim - 0.5)
-    ax.set_ylim(-0.5 + y_trim, image_height - y_trim - 0.5)
+    ax.set_xlim(-0.5 + bbox[0], bbox[1] - 0.5)
+    ax.set_ylim(-0.5 + bbox[2], bbox[3] - 0.5)
```
```suggestion
    ax.set_xlim(-0.5 + bbox[0], bbox[1] + 0.5)
    ax.set_ylim(-0.5 + bbox[2], bbox[3] + 0.5)
```
Because the bbox calculation gives a closed range e.g. if all pixels are finite then bbox[1] will be `image_width - 1` rather than `image_width`.

mauch commented:
@bmerry I've added an option to change the facecolor of the plots and fixed up the default bounding box range. Can you please have one more look.

At the moment the tests of png output just involve checking that an image is produced. This could possibly be expanded but will need some way of reading png output into an array to check the pixel values (pillow probably will do). I think any further tests can wait until this ends up in its own repository.

bmerry commented:
> At the moment the tests of png output just involve checking that an image is produced.

I think that's fine. It really needs a human to look at them to see if they're "right" e.g. text is all readable, doesn't overlap, sane colour scheme etc.

I'd suggest we merge this; moving the code to a common place can happen afterwards.
